beacon-chain/rpc/blockfetcher/fetcher.go

```diff
@@ -0,0 +1,101 @@
+package blockfetcher
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/blockchain"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/db"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/interfaces"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+)
+
+// BlockIdParseError represents an error scenario where a block ID could not be parsed.
+type BlockIdParseError struct {
+	message string
+}
+
+// NewBlockIdParseError creates a new error instance.
+func NewBlockIdParseError(reason error) BlockIdParseError {
+	return BlockIdParseError{
+		message: errors.Wrapf(reason, "could not parse block ID").Error(),
+	}
+}
+
+// Error returns the underlying error message.
+func (e BlockIdParseError) Error() string {
+	return e.message
+}
+
+// Fetcher is responsible for retrieving info related with the beacon block.
+type Fetcher interface {
```

Comment:
Sort of a nitpick - I think the naming could be improved, and pulling on that string I want to suggest a different package layout.

Start with - the name `blockfetcher.Fetcher` stutters. Conventional wisdom, particularly in the go world, is that this kind of naming adds overhead to reading code by introducing redundant information. Fetcher also breaks the guidance of naming an interface after its methods.  Between the struct and the interface, we have the words "Provider" and "Fetcher", which I think were only chosen as synonyms for "get a thing" because the struct and the interface need different names.

For the concrete type, it's nice when the name takes the interface type and moves it closer to the concrete implementation by describing *how* it's satisfying the interface. So for instance assume we name the interface for the `Block` method `Blocker`, the implementation that uses `BeaconDB` could be `BeaconDBBlocker. 

I think all of these "fetcher" types could make sense in the same package, as they are sort of wrappers around database + blockchain package (disguisded as "chaininfo"). I'm not sold on that idea but I think it will make reading code a bit more intuitive and bring some related types closer together. What if we put them all in a single package for looking up core consensus values? eg `lookup.Blocker` being an interface that has the method `Block(...)`. `lookup.Stater` has `State(...)` and so on. Lookup is just a strawman, other possibilities could be `retrieve` or `load`. 

I'm adding a new package name verb to the mix here because I feel like "fetch" and "provide" have baggage in the broader world of software. "Fetch" often implies getting something from *somewhere else* - you don't generally ask someone to "fetch" something directly at hand, usually it means something like making a network request. "Provide" is often used in the programming language sense of a concrete implementation for an interface (in the same sense that go uses "satisfy"). So in context it isn't clear to me if `BlockProvider` is supposed to "provide" the interface that "Fetcher" defines (but then why does it use the word "Block" and not the word "Fetcher" in its name?), or if "provider" is being used to say "it knows how to *provide* this sort of value given this identifier". 

I wouldn't block the PR over this suggestion but I do think we should improve the naming here before we add more "fetcher" types.

Reply:
> Between the struct and the interface, we have the words "Provider" and "Fetcher", which I think were only chosen as synonyms for "get a thing" because the struct and the interface need different names.

So true :sweat_smile: 

I really appreciate this criticism. Everything you say makes perfect sense and I am going to adopt this.